vsintegration/src/FSharp.Editor/CodeLens/FSharpCodeLensService.fs

```diff
@@ -191,37 +198,37 @@ type internal FSharpCodeLensService
 #endif
                                 return None
                         else return None
-                    with e -> 
+                    with e ->
 #if DEBUG
                         logErrorf "Error in lazy line lens computation. %A" e
 #else
                         ignore e
 #endif
                         return None
                 }
-            
+                
             let inline setNewResultsAndWarnIfOverriden fullDeclarationText value = 
 #if DEBUG
                 if newResults.ContainsKey fullDeclarationText then
                     logWarningf "New results already contains: %A" fullDeclarationText
 #endif
                 newResults.[fullDeclarationText] <- value
-
+                
             for symbolUse in symbolUses do
                 if symbolUse.IsFromDefinition then
                     match symbolUse.Symbol with
                     | :? FSharpMemberOrFunctionOrValue as func when func.IsModuleValueOrMember || func.IsProperty ->
-                        let funcID = func.FullName
-                        let fullDeclarationText = funcID
+                        let funcID = func.LogicalName + (func.FullType.ToString() |> hash |> string)
+                        // Use a combination of the the function name + the hashed value of the type signature
                         let fullTypeSignature = func.FullType.ToString()
                         // Try to re-use the last results
-                        if lastResults.ContainsKey fullDeclarationText then
+                        if lastResults.ContainsKey funcID then
```

Comment:
Since you're hashing the text, should `lastResults` have its key be an `int` rather than a `string`? This would avoid an extra allocation.

Reply:
If we want to minimize allocations I need to look into this again.

Maybe it would be enough to make a dictionary which maps from function name to the hash of the full type (last time it was used) and then only compare the value/hash of the last type to change only the code lens above functions which type has really changed. Would that allocate less?

Reply:
Generally no, it's preferable not to use a string as a key if you can get away with it. Although it's still all small allocations, if you're already hashing something then that would be adequate for a key.

Though generally it's recommended to use more information than just a hash for a key, since there can be hash collisions.

Alternatively, does this caching provide any notable benefits? That is, if nothing was cached, does it meaningfully affect UX when used?

Reply:
Of course, I wouldn't apply it if it would be of benefit :D
The current problem is that the key isn't unqiue which results in code lenses which aren't removed.

I think about removing the `|> hash` as it's better to hash the whole string anyways (so no hash collisions, thanks for the reminder btw). Currently I cannot see any possibility to remove the usage of a string as key. The small allocation is still much better than what would result in not using the cache (because that's what you get if you don't want it :D).